workflows/mapping/Snakefile

```diff
@@ -249,4 +257,67 @@ rule merge_bam:
     wrapper:
         wrapper_for('samtools/merge')
 
-# vim: ft=python
+
+rule fastq_count:
+    input: '{prefix}.fastq.gz'
+    output: '{prefix}.fastq.count'
+    shell: "gunzip -c {input[0]} | echo $((`wc -l` / 4)) > {output[0]}"
+
+
+rule bam_count:
+    input: '{prefix}.bam'
+    output: '{prefix}.bam.count'
+    shell: "samtools view -c {input[0]} > {output[0]}"
+
+rule sample_count_table:
+    input: fq = '{prefix}.fastq.count',
+           cutadapt = '{prefix}.cutadapt.fastq.count',
+           align = '{prefix}.cutadapt.ht2.bam.count',
+           dedup = '{prefix}.cutadapt.ht2.unique.sort.dedup.bam.count'
+    output: '{prefix}_counts.csv',
+    run:
+        import pandas as pd
+        def read_count(count):
+            with open(count, 'r') as FH:
+                return int(FH.read().strip())
+        
+        fq = read_count(input.fq)
+        cutadapt = read_count(input.cutadapt)
+        align = read_count(input.align)
+        dedup = read_count(input.dedup)
+
+        df = pd.DataFrame([[wildcards.prefix, fq, cutadapt, align, dedup]], columns=['sampleID', 'Number Reads', 'Number Trimmed Reads', 'Number Aligned', 'Number Uniquely Aligned'])
+        df.to_csv(output[0], index=False)
+
+rule count_table:
+    input: SH.build_targets(['{runLevel}_counts.csv'])
+    output: csv = 'count_table.csv',
+            html = 'count_table.html'
+    run:
+        import pandas as pd
+
+        dfs = [pd.read_csv(x) for x in input]
+        df = pd.concat(dfs)
+        df.to_csv(output.csv, index=False)
+        with open(output.html, 'w') as OUT:
+            df.to_html(buf=OUT, index=False, classes='table table-striped')
+
+rule report:
+    input: Table1='count_table.html'
+    output: html='report.html'
+    run:
+        from reporting import report
+
+        myIn = {}
+        with open(input.Table1, 'r') as FH:
+            myIn['Table1'] = FH.read().strip()
+        
```

Comment:
Agreed. `input:` can be a dict of all the chunks created so far.

Using the jinja-chunk-iterator template above, the following might work. If you're generating HTML for each chunk, you can bypass markdown and all the annoying indentation issues. The idea is that the jinja templates do all the work. In fact the conditionals on whether or not chunk X should be included should be in the jinja template (detecting whether or not they got a context item for, say, trimming chunks):

Say we have this in `templates/base.html`:

```jinja
<div id="content">
    {% for chunk in chunks %}
    <div id="{{chunk.id}}">
        {{ chunk.html }}
    </div>
    {% endfor %}
</div>
```

Then:

```python

class Chunk(object):
    def __init__(self, name, html_fn):
        self.id = name
        self.html = open(html_fn).read()

chunks = [Chunk(*i) for i in input.items()]       

# in reporting.report, provide chunks to template;
# let jinja do the work

def report(template, **kwargs):
    tp = ENV.get_template('base.html')
    tp.render(**kwargs)


report(chunks=chunks)

```